core/trino-main/src/test/java/io/trino/operator/window/matcher/TestPatternMatchingMachine.java

```diff
@@ -0,0 +1,215 @@
+import static io.trino.sql.planner.rowpattern.TestUtils.concatenation;
+import static io.trino.sql.planner.rowpattern.TestUtils.empty;
+import static io.trino.sql.planner.rowpattern.TestUtils.end;
+import static io.trino.sql.planner.rowpattern.TestUtils.excluded;
+import static io.trino.sql.planner.rowpattern.TestUtils.label;
+import static io.trino.sql.planner.rowpattern.TestUtils.permutation;
+import static io.trino.sql.planner.rowpattern.TestUtils.plusQuantified;
+import static io.trino.sql.planner.rowpattern.TestUtils.questionMarkQuantified;
+import static io.trino.sql.planner.rowpattern.TestUtils.starQuantified;
+import static io.trino.sql.planner.rowpattern.TestUtils.start;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.lang.String.format;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class TestPatternMatchingMachine
+{
+    private static final Map<IrLabel, Integer> LABEL_MAPPING = ImmutableMap.of(
+            new IrLabel("A"), 0,
+            new IrLabel("B"), 1,
+            new IrLabel("C"), 2,
+            new IrLabel("D"), 3,
+            new IrLabel("E"), 4);
+
+    @Test
+    public void testCaptures()
+    {
+        assertCaptures(
+                concatenation(excluded(label("A")), label("B")),
+                "ABCD",
+                new int[] {0, 1});
+
+        assertCaptures(
+                excluded(concatenation(label("A"), label("B"))),
+                "ABCD",
+                new int[] {0, 2});
+
+        assertCaptures(
+                concatenation(label("A"), excluded(label("B"))),
+                "ABCD",
+                new int[] {1, 2});
+
+        assertCaptures(
+                concatenation(label("A"), starQuantified(excluded(label("B")), true)),
+                "ABBBCD",
+                new int[] {1, 2, 2, 3, 3, 4});
+
+        assertCaptures(
+                concatenation(label("A"), excluded(starQuantified(label("B"), true))),
+                "ABBBCD",
+                new int[] {1, 4});
+
+        assertCaptures(
+                concatenation(label("A"), starQuantified(excluded(label("B")), true), label("C"), excluded(starQuantified(label("D"), true))),
+                "ABBCDDDE",
+                new int[] {1, 2, 2, 3, 4, 7});
+
+        assertCaptures(
+                concatenation(label("A"), starQuantified(concatenation(excluded(concatenation(label("B"), label("C"))), label("D"), excluded(label("E"))), true)),
+                "ABCDEBCDEBCDE",
+                new int[] {1, 3, 4, 5, 5, 7, 8, 9, 9, 11, 12, 13});
+    }
+
+    @Test
+    public void testMatchWithoutCaptures()
+    {
+        assertCaptures(concatenation(label("A"), label("B")), "ABCD", new int[] {});
+    }
+
+    @Test
+    public void testEmptyMatch()
+    {
+        assertCaptures(empty(), "ABCD", new int[] {});
+        assertLabels(empty(), "ABCD", new char[] {});
+    }
```

Comment:
We should be able to do something similar to ParserAssert or QueryAssert. Then you might write this as:

```
assertThat(match(empty(), "ABCD"))
    .hasCaptures(...)
    .hasLabels(...)
```